Navigation: preload more API requests

anton-vlasenko opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
When loading the Navigation screen, there are a lot of loading spinners due to API requests to /menus and /menu-items. We should preload these requests so that the screen appears faster.
The goal of this PR is to preload these requests.
Also, this PR fixes 2 small bugs in the `core-data` and `api-fetch` modules.
Fixes https://github.com/WordPress/gutenberg/issues/24642

## How has this been tested?
1. Open the navigation page (`wp-admin/admin.php?page=gutenberg-navigation`).
2. Make sure there is only 1 fetch/XHR request (see the screenshot below).

## Screenshots <!-- if applicable -->
Before:
![Screenshot 2021-10-06 at 16 44 50](https://user-images.githubusercontent.com/43744263/136235248-16a0f1ae-0324-4e45-98dd-b7b0d155b288.png)
After:
![Screenshot 2021-10-06 at 17 29 11](https://user-images.githubusercontent.com/43744263/136235260-b53acf9c-a0cd-4070-8881-ddeee5c17945.png)

## Types of changes
New feature
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

anton-vlasenko committed:
Add preloading for /__experimental/menus endpoint.

anton-vlasenko committed:
Add a new filter to allow to modify the data that has already been preloaded.

anton-vlasenko committed:
Commit WIP.

anton-vlasenko committed:
Preload more endpoints.

anton-vlasenko committed:
Fix preload paths.

anton-vlasenko committed:
Don't add slash if the key is empty.
Otherwise we will get urls like /wp/v2/types/?context=edit (trailing slash).

anton-vlasenko committed:
Preload /wp/v2/types URLs.

anton-vlasenko committed:
Commit WIP.

anton-vlasenko committed:
1. Fix the test.
2. Add PHPDOC blocks.
3. Remove {$editor_name}_preloaded_data filter. We are not going to use it.

anton-vlasenko committed:
Fix indents.

anton-vlasenko committed:
Add PHPDOC blocks to the public functions introduced in this PR.

anton-vlasenko committed:
Add @since tags.

anton-vlasenko committed:
OPTIONS type requests should be cached the same way as GET type requets.
We must allow using cached data only once (to avoid using outdated data).

anton-vlasenko commented:
We still have 1 XHR request left because we can only use data from the cache once.
Then, to avoid loading stale (outdated) data, the data gets deleted from the cache.
Please see [this](https://github.com/WordPress/gutenberg/blob/trunk/packages/api-fetch/src/middlewares/preloading.js#L56) for reference.
So, when we fetch the data for the second time, we perform an actual fetch request (because we've removed the data from the cache).
I partially agree with [this comment](https://github.com/WordPress/gutenberg/pull/24672#issuecomment-680375627), and IMO, the proper solution to this would be to cache data at the `wordpress/data` level.

anton-vlasenko committed:
Fix the code style.

TimothyBJacobs commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,101 @@ class="edit-navigation"
+ * @return string
+ */
+function gutenberg_navigation_get_menus_endpoint( $results_per_page = 100 ) {
+	return '/__experimental/menus?' . http_build_query(
```
Should this be using `build_query`?

anton-vlasenko replied:
Yes, it should. Good catch.
Fixed in ccb369243cdf2840175f2fd7fc29d36b181870ec

anton-vlasenko committed:
Use WP function build_query instead of http_build_query.

anton-vlasenko committed:
Fix code style.

noisysocks commented on packages/api-fetch/src/middlewares/preloading.js:
```diff
@@ -72,11 +72,11 @@ function createPreloadingMiddleware( preloadedData ) {
-				);
+				// Unsetting the cache key ensures that the data is only preloaded a single time
+				const cacheData = cache[ method ][ path ];
+				delete cache[ method ][ path ];
```
Tiny nit: would be nice to format this similarly to the above branch

```suggestion
				const cacheData = cache[ method ][ path ];
				
				// Unsetting the cache key ensures that the data is only preloaded a single time
				delete cache[ method ][ path ];
				
```

anton-vlasenko replied:
I agree. 
Fixed in d71c48f4cf4c51da33a598e5381f21cd07d8eab9

noisysocks commented:
> So, when we fetch the data for the second time, we perform an actual fetch request (because we've removed the data from the cache).

Why are we fetching data a second time?

noisysocks reviewed (approved):
Code LGTM and I can see only one XHR request in DevTools when I browse to the navigation screen 👍

spacedmonkey commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,101 @@ class="edit-navigation"
+		array(
+			'per_page' => $results_per_page,
+			'context'  => 'edit',
+			'_locale'  => 'user',
```
Is this needed?

anton-vlasenko replied:
Yes, this is needed.
Cache records are matched by the path.
If we don't use this path, we will not be able to find the cache record.

adamziel replied:
I don't like how the path must be kept in sync between two separate places (JS and PHP), but I also don't have any better ideas to offer. 😒

spacedmonkey commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,101 @@ class="edit-navigation"
+		gutenberg_navigation_get_types_endpoint(),
 	);
 
+	if ( $first_menu_id ) {
```
Is it always the first menu id? I thought it loaded from the selected menu id.

spacedmonkey replied:
Look at this. The currently selected menu is used in user storage. So couldn't a user select a different menu? Then reload the page, meaning that a different menu needs to be reloaded here? Is there anyway to this selected menu id to php? 

![image](https://user-images.githubusercontent.com/237508/136600589-2d50333e-fd36-407d-b2a8-b179bd45d201.png)

anton-vlasenko replied:
Yes, I'm aware of this.
There is no way to pass it to PHP unless we want to make an additional REST API request every time a user changes the currently selected menu.
I've implemented a middleware that fixes the issue with the $first_menu_id though.

spacedmonkey replied:
CC @anton-vlasenko

spacedmonkey commented:
How does this handle the fact the REST API preloaded, means that caches have to be bust. See https://github.com/WordPress/gutenberg/issues/24642#issuecomment-908724768

anton-vlasenko committed:
Format the code.

anton-vlasenko commented:
@spacedmonkey 
> How does this handle the fact the REST API preloaded, means that caches have to be bust. See [#24642 (comment)](https://github.com/WordPress/gutenberg/issues/24642#issuecomment-908724768)

Thank you for the review.
To reply to your comment, I need to understand what you mean by "caches" when you say they have to be bust.
Do you mean browser cache or some other type of cache?

anton-vlasenko committed:
Add a new filter to modify the preloaded data.
This can be handy in many situations when we need to change the preloaded data.

spacedmonkey commented:
By cache, I mean the preloading itself. That is a cache, it doesn't matter how many times you reload the api request, it will still come from the preloaded middleware. My worry, is that even if we update the menu items, that the data will not update, as it is cache. 

By cache busting, i.e just add a query string to the api request, that will stop the request loading from middleware.

anton-vlasenko committed:
Wire up createMenuPreloadingMiddleware so that we can preload menus data separately.

anton-vlasenko committed:
Fix PHPDOC blocks.

anton-vlasenko committed:
Add PHPDOC block.

spacedmonkey commented on packages/api-fetch/src/middlewares/preloading.js:
```diff
@@ -72,11 +72,12 @@ function createPreloadingMiddleware( preloadedData ) {
+				const cacheData = cache[ method ][ path ];
+
+				// Unsetting the cache key ensures that the data is only preloaded a single time
+				delete cache[ method ][ path ];
```
This is a very big change to how preloading works. Are we saying here, after the first load, preload cache is removed. What about other problems of the code that may call the same rest api twice. 

Cc @swissspidy

swissspidy replied:
This is not new/changed behavior? See https://github.com/WordPress/gutenberg/blob/0c5d68755496e56d5558e47d063f31d97f05d8c3/packages/api-fetch/src/middlewares/preloading.js#L55

spacedmonkey replied:
How is not a new behavior?

swissspidy replied:
That‘s from a few lines above. Sounds like it was not fully implemented there.

so this change here looks reasonable 👍

anton-vlasenko replied:
> That is a cache, it doesn't matter how many times you reload the api request, it will still come from the preloaded middleware.

This is not how it works at the moment.
We delete the cache data from the preloadedData object the first time  we use the data, so nothing gets cached and we don't need to bust the caches.

anton-vlasenko replied:
> What about other problems of the code that may call the same rest api twice.

Then we need to fix those parts of the code that call the same REST API twice. We should call the same REST API twice only if it's absolutely needed.
IMO preloading cannot fix everything.

swissspidy replied:
Yeah, this here updates the cache logic for `OPTIONS` requests to match the behavior for `GET` requests. All reasonable to me.

getdave replied:
Annoying as it is we should consider pulling this out into a separate PR. That way it will be easier to notice (and rollback) if this introduces any unforeseen bugs.

spacedmonkey replied:
> That way it will be easier to notice (and rollback) if this introduces any unforeseen bugs.

Other packages and scripts may use this package, this change of behaviour may break something. This needs to be a changelog and well documented.

anton-vlasenko replied:
I've moved this code into a separate PR.
https://github.com/WordPress/gutenberg/pull/35527
We should be good to merge now.

spacedmonkey commented:
<img width="1884" alt="Screenshot 2021-10-10 at 00 19 57" src="https://user-images.githubusercontent.com/237508/136676080-73461ebd-04cd-4052-bc8c-97ea30d8d08a.png">
I see this error when I checkout this branch and build it. 

Also I am seeing PHPUnit tests failing.

swissspidy commented on lib/editor-settings.php:
```diff
@@ -75,6 +75,14 @@ function gutenberg_initialize_editor( $editor_name, $editor_script_handle, $sett
+	/**
+	 * Filters the array of data that has been preloaded.
+	 *
+	 * @param Array $preload_data Array containing the preloaded data.
```
```suggestion
	 *
	 * The dynamic portion of the hook name, `$editor_name`, refers to the type of block editor.
	 * 
	 * @param Array $preload_data Array containing the preloaded data.
```

swissspidy commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,100 @@ class="edit-navigation"
+ * @param  int $menu_id Menu ID.
+ * @return string
+ */
+function gutenberg_navigation_get_menu_endpoint( $menu_id ) {
```
Where are these functions used?

anton-vlasenko replied:
Nowhere :) I was going to delete it. Good catch.

swissspidy commented on packages/api-fetch/src/middlewares/preloading.js:
```diff
@@ -72,11 +72,12 @@ function createPreloadingMiddleware( preloadedData ) {
+				const cacheData = cache[ method ][ path ];
+
+				// Unsetting the cache key ensures that the data is only preloaded a single time
+				delete cache[ method ][ path ];
```
Should this be mentioned in the changelog? Bummer that #25550 didn't do that originally either

anton-vlasenko replied:
Yes, it should. I will move this fix to a new PR.

anton-vlasenko committed:
Implement menu preloading middleware.

anton-vlasenko committed:
Remove the gutenberg_navigation_get_menu_endpoint function. It's not used.

getdave commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -16,3 +16,106 @@ export function stripHTML( html ) {
 	return document.body.textContent || '';
 }
+
+/**
```
Could we document this function with 

1. Why it exists.
2. High level overview of what it is doing.

🙏

anton-vlasenko replied:
Fixed in https://github.com/WordPress/gutenberg/pull/35402/commits/24db010503e5f9bb37003b7208db9b1854bfcc0a

anton-vlasenko committed:
We should only load data once.

anton-vlasenko committed:
Add PHPDOC block.

anton-vlasenko committed:
Fix the unit test.

anton-vlasenko committed:
Improve test code coverage. Add a new unit test for test_gutenberg_navigation_editor_preload_menus_function_returns_correct_data

anton-vlasenko committed:
Update lib/editor-settings.php

Co-authored-by: Pascal Birchler <pascalb@google.com>

anton-vlasenko committed:
Fix code style.

anton-vlasenko commented:
@spacedmonkey 

> <img alt="Screenshot 2021-10-10 at 00 19 57" width="1884" src="https://user-images.githubusercontent.com/237508/136676080-73461ebd-04cd-4052-bc8c-97ea30d8d08a.png">
> 
> I see this error when I checkout this branch and build it.
> Also I am seeing PHPUnit tests failing.

I can't reproduce that console error. Are you sure you executed `npm install && npm run dev`?
Unit tests should be fine now.

anton-vlasenko committed:
Fix code style.

anton-vlasenko committed:
Fix code style.

spacedmonkey commented:
I am now seeing menu locations / menus requested, even through this data is preloaded. 

#### Trunk

<img width="1904" alt="Screenshot 2021-10-11 at 17 57 37" src="https://user-images.githubusercontent.com/237508/136827896-182f248c-c07d-4ed0-8a52-0b4b678678c3.png">

#### This PR
<img width="1904" alt="Screenshot 2021-10-11 at 17 55 05" src="https://user-images.githubusercontent.com/237508/136827951-eb75b416-8d24-44b4-a224-f2f421719865.png">

So now in this branch, we are seeing, 
- Menu locations
- Menus 
- First menu REST API requests. 

All of this data is preloaded and should not be loaded again. 

This is related to this thread https://github.com/WordPress/gutenberg/pull/35402/files#r725465023

After removing `delete cache[ method ][ path ];`, This is what the requests looked like

<img width="1884" alt="Screenshot 2021-10-11 at 18 08 46" src="https://user-images.githubusercontent.com/237508/136828861-69687c37-1fa8-46ea-8365-31c0ea255364.png">

Let's remove this `delete cache[ method ][ path ];` for now as it is having unexpected side effects.

anton-vlasenko commented:
@spacedmonkey 

> Let's remove this `delete cache[ method ][ path ];` for now as it is having unexpected side effects.

This is a side effect of the new additional middleware.
According to my tests it has nothing to do with `delete cache[ method ][ path ];`.
And it shouldn't happen just because we've added a new middleware.
Let me quickly investigate this.

anton-vlasenko committed:
Fix bug in createMenuPreloadingMiddleware.
We should send correct response for /__experimental/menu/<menuId> endpoint.

anton-vlasenko reviewed (commented):
> Let's remove this `delete cache[ method ][ path ];` for now as it is having unexpected side effects.

I've found a bug in the middleware, so this should be fixed now.
Thank you for testing.
Fixed in  fcfdc8c0bbef408ebd2b373ed92c5cbddab58de1

anton-vlasenko committed:
Revert this change as it needs to be moved into a separate PR.

anton-vlasenko committed:
Add public access modifier to the functions.

spacedmonkey commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,84 @@ class="edit-navigation"
 	}
 
+	$menus         = wp_get_nav_menus();
+	$first_menu_id = count( $menus ) ? $menus[0]->term_id : null;
```
```suggestion
	$first_menu_id = ! empty( $menus ) ? $menus[0]->term_id : null;
```

spacedmonkey reviewed (approved):
Approved with a nicpic 🎉

adamziel commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,84 @@ class="edit-navigation"
+/**
+ * This function returns an url for the /wp/v2/types endpoint
+ *
+ * @since 11.6.0
```
It's going to be 11.8 now

anton-vlasenko replied:
Fixed in 83a1edc937479edd1df738b322f7e79df0d9cdb4

adamziel commented on lib/navigation-page.php:
```diff
@@ -82,3 +143,38 @@ function gutenberg_navigation_editor_load_block_editor_scripts_and_styles( $is_b
+			:
+			array(
+				$menus_data_path => $preload_data[ $menus_data_path ],
+			);
```
That's such a nit pick, but I find this ternary formatting pretty hard to read. Perhaps an `if/else` would be more readable?

anton-vlasenko replied:
I agree. It looks ugly.
I've fixed this, and it seems we don't need `if/else` here, `if` is enough.

adamziel commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,128 @@
+ * It uses data that is returned from the /__experimental/menus endpoint for requests
+ * to the /__experimental/menu/<menuId> endpoint, because the data is the same.
+ * This way, we can avoid making additional REST API requests.
+ * This middleware can be removed if/when we implement caching at the wordpress/data level.
```
Would you create a ticket for caching at the wordpress/data level that so that we don't forget about it?

anton-vlasenko replied:
Yes, https://github.com/WordPress/gutenberg/issues/35601

adamziel commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,128 @@
+		if ( ! key ) {
+			return next( options );
+		}
+
```
Is this check useful here?

```suggestion
```

anton-vlasenko replied:
It's better safe than sorry, but I agree, we don't need that.

adamziel commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,128 @@
+		const menuId = parseInt( matches[ 1 ] );
+		const menu = menuData.filter( ( { id } ) => {
+			return id === menuId;
+		} );
```
```suggestion
		const menu = menuData.filter( ( { id } ) => id === menuId );
```

adamziel commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,128 @@
+			return id === menuId;
+		} );
+
+		if ( 0 < menu.length ) {
```
Yoda conditions don't read well, e.g. here it's "if zero is less than menu length then." I know there are some guidelines to do that on PHP side to prevent accidental assignments like `$a = 4` instead of `$a === 5`, but these days linters are pretty good at catching that (and also it doesn't apply to `<`).

```suggestion
		if ( menu.length ) {
```

anton-vlasenko replied:
> Yoda conditions don't read well,

I see what you mean. I can agree it can be harder to read Yoda conditions for some people.
Although Yoda conditions have their advantages and it's not a crime to use them. IMO it's just a personal preference.

`if ( value > 0 )` is better than just `if ( value )`, because the former is a stricter comparison.
If `value ===  -1`, the latter expression will return `true`, which is not what it is supposed to do.
This can happen if we compare some unsigned numbers like bank account amounts.
I understand that length of the array cannot be lower than zero. But I prefer to write it like that everywhere because it will work correctly regardless of whether we compare unsigned or signed integers.

anton-vlasenko replied:
Fixed (?) in 612d17b6d1a07a83db7bf35173efddc14e1b9897

adamziel commented:
My concern here is that we'll soon end up with `createWidgetsPreloadingMiddleware`, `createPostsPreloadingMiddleware` and so on. Reusing data from the list endpoint in another endpoint sounds like a common use case – how is this solved in different places? If it isn't, are there any other places that could use such a mechanism? If so, could this be incorporated into the existing `createPreloadingMiddleware`

TimothyBJacobs commented on lib/editor-settings.php:
```diff
@@ -75,6 +75,16 @@ function gutenberg_initialize_editor( $editor_name, $editor_script_handle, $sett
+	 *
+	 * @param Array $preload_data Array containing the preloaded data.
+	 */
+	$preload_data = apply_filters( "{$editor_name}_preload_data", $preload_data );
```
Is this a new filter? We should try to avoid using dynamic filter names.

TimothyBJacobs replied:
Hm, isn't this already filterable in `block_editor_rest_api_preload` actually?

anton-vlasenko replied:
Yes, this is a new filter.
> We should try to avoid using dynamic filter names.

Good to know.
Fixed in 75519d9b4e4c08585f912d234c79041e063b77a7

TimothyBJacobs replied:
Why do we need this filter then? Why isn't it covered by `block_editor_rest_api_preload`?

anton-vlasenko replied:
@TimothyBJacobs 
It isn't covered by `block_editor_rest_api_preload` function because we need to modify the data that gets passed into the `'wp.apiFetch.use( wp.apiFetch.createPreloadingMiddleware( %s ) );'`.
My bad, I didn't notice you mentioned `block_editor_rest_api_preload` the first time.
I can refactor this to use `block_editor_rest_api_preload` instead, but:
1.  I will have to add a new filter into the body of the `block_editor_rest_api_preload` function because of the reason above. 
2. We will break backward compatibility because we will have to get rid of the
```php
$preload_paths = apply_filters( "{$editor_name}_preload_paths", $settings['preload_paths'] );
```
call.
The `block_editor_rest_api_preload` function will replace all that code.
We need to refactor this, but I'm not sure about breaking BC. It seems that we don't use the filter above (`{$editor_name}_preload_paths`).
What are your thoughts on this?

TimothyBJacobs replied:
What I'm having trouble understanding is why we can't use the `block_editor_rest_api_preload_paths` filter in `block_editor_rest_api_preload`. Why do we need another filter on a list of preloaded paths.

anton-vlasenko replied:
@TimothyBJacobs 

This is not a filter on a list of preloaded paths.
This is the filter that allows modifying already preloaded data.
I've created `createMenuPreloadingMiddleware` middleware, and it needs its own set of data, i.e., a list of all menus.
That filter allows modifying preloaded data so that list of all menus doesn't get passed into the `createPreloadingMiddleware`.
 Instead, we need to pass it into the `createMenuPreloadingMiddleware` middleware and process it there.
We have to either push this PR with that additional filter OR remove that filter and the custom middleware (`createMenuPreloadingMiddleware`).
In such a case, we won't be able to preload `/__experimenta/menu/<menuId>` endpoints properly until we find a better solution.
What is your opinion on this?

anton-vlasenko committed:
Fix version name.

anton-vlasenko committed:
Update packages/edit-navigation/src/utils/index.js

Co-authored-by: Adam Zielinski <adam@adamziel.com>

anton-vlasenko committed:
Remove the yoda condition as it can be harder to read it for some people.

anton-vlasenko committed:
1. Format the code.
2. Don't use variable filter name. Pass the ditor_name as the second parameter.

anton-vlasenko committed:
1. Update PHPDOC block.
2. Remove redundant code.

anton-vlasenko committed:
Update lib/navigation-page.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

anton-vlasenko commented:
@noisysocks 

> > So, when we fetch the data for the second time, we perform an actual fetch request (because we've removed the data from the cache).
> 
> Why are we fetching data a second time?

It's hard to tell why, but both requests are coming from this place: https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/resolvers.js#L105
I've created an issue so that we don't forget about it.
https://github.com/WordPress/gutenberg/issues/35601

anton-vlasenko committed:
Fix the comment.

talldan commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,122 @@
+ *
+ * @return {string} Normalized path.
+ */
+export function getStablePath( path ) {
```
Looking at the way this function works, it's worth checking some of the utils in the `url` package. It might be possible to make use of `getQueryString` or `getQueryArgs` from that package.

Some of this code might also be a good candidate for the `url` package. I could see a `normalizeQueryString` function extracted from this being a useful utility. There might also be overlap with `buildQueryString` does that already order query string args?

Might be a good thing for a separate follow-up PR.

anton-vlasenko replied:
Awesome! I've created a GH issue for this. https://github.com/WordPress/gutenberg/issues/35799

talldan commented on packages/edit-navigation/src/utils/index.js:
```diff
@@ -0,0 +1,122 @@
+ * @param {Object} preloadedData
+ * @return {Function} Preloading middleware.
+ */
+export function createMenuPreloadingMiddleware( preloadedData ) {
```
Even though `edit-navigation` isn't a public package, I'm uneasy about exporting this as a stable API because it's very specific to menus and the navigation editor.

It seems like a more general problem that when we preload a request that lists items the individual item can also be considered preloaded, regardless of the REST resource. It might be something that applies to posts, pages and other resources too. What do you think?

I'm not against merging this PR with a special case for menus right now, but in the long run I could see this as a candidate for removal if we come up with a more generalised preloading system.

Right now it might be worth exporting this as `__unstableCreateMenuPreloadingMiddleware` so that it doesn't accidentally end up as a stable API that has to be supported indefinitely.

anton-vlasenko replied:
I'd wait until https://github.com/WordPress/gutenberg/issues/15105 gets merged and remove `createMenuPreloadingMIddleware` altogether.
When I started to work on this issue I wanted to implement something like that. I didn't know this PR already existed.
Thank you for the review.

adamziel replied:
#15105 is an issue and seems unrelated. Did you mean another number?

talldan replied:
@adamziel The related discussion is here - https://github.com/WordPress/gutenberg/pull/35402#discussion_r728639376, the background is that we'd be able to preload the selected menu item if it were stored in the database instead of local storage, and I guess that removes the use case for this function? Though I still feel like we'll be in a situation where we need to load all the menus for the menu switcher.

@anton-vlasenko I wouldn't worry too much about waiting for #15105 / #19177 because it has been a very long-running issue, but it'd be good to help it along if we can.

If we can ship this PR by renaming the function I think that's fine 👍

anton-vlasenko replied:
@adamziel 
> #15105 is an issue and seems unrelated. Did you mean another number?

It seems to be related. **If** it allows to store `selectedMenuId` to the database and then fetch it from there, it will solve all the issues with preloading.
We have to tinker with `createMenuPreloadingMiddleware` only because we don't know which menu is currently selected when preloading data.

anton-vlasenko replied:
Thank you for the feedback, @talldan.
I will surely rename that method. (UPD: fixed in 41ea18345ef522b25ec4ebfc9d2ba6a7dbd50a2e).
I'm waiting for @TimothyBJacobs 's feedback on this comment. https://github.com/WordPress/gutenberg/pull/35402#discussion_r729886999

talldan commented on lib/navigation-page.php:
```diff
@@ -21,23 +21,84 @@ class="edit-navigation"
 	);
 
+	if ( $first_menu_id ) {
+		$preload_paths[] = gutenberg_navigation_get_menu_items_endpoint( $first_menu_id );
```
I see there was already a resolved conversation about this, but I think it might be worth creating an issue to explore how the selected menu might be preloaded.

A thought I had is that we could pass the menu id as a query string `http://localhost:8888/wp-admin/nav-menus.php?menuId=2`, but it'd require every link to the editor to have that query string 🤯 . Deep linking to a menu would be cool though. A query string wouldn't be perfect, even better would be a proper route like `http://localhost:8888/wp-admin/menus/2`.

Or alternatively we could help with https://github.com/WordPress/gutenberg/issues/15105, at which point the preference would be stored in the database rather than localStorage, and our server-side code would be able to use the value.

anton-vlasenko replied:
> Or alternatively we could help with #15105, at which point the preference would be stored in the database rather than localStorage, and our server-side code would be able to use the value.

That PR will be a game-changer if it works as you described. Then we could just use `selectedMenuId` from the database to preload data and get rid of the createMenuPreloadingMiddleware altogether.

anton-vlasenko commented:
@adamziel 
> > My concern here is that we'll soon end up with `createWidgetsPreloadingMiddleware`, `createPostsPreloadingMiddleware` and so on. Reusing data from the list endpoint in another endpoint sounds like a common use case – how is this solved in different places?

I'm not aware of how this is solved in different places.
I haven't been able to find any examples.

> If it isn't, are there any other places that could use such a mechanism? If so, could this be incorporated into the existing `createPreloadingMiddleware`

I've tried to incorporate this into the `createPreloadingMiddleware` via callback.
I implemented a callback that allows modifying preloaded data. But,  there were issues with stale data if a user loads the Navigation Editor page and tries to switch to another menu.
It's because we need to use menu data only once. Unfortunately, callbacks don't allow us to delete that data after it was used.
I find it difficult to explain, but callbacks are not likely going to work.

anton-vlasenko committed:
Make the middleware unstable.
We don't want it to become a stable API for now.
See https://github.com/WordPress/gutenberg/pull/35402#discussion_r728636852

TimothyBJacobs commented:
The structure of this PR is confusing to me. I don't understand why we are introducing another filter in the block editor settings. I understand that this is filtering the preloaded data, instead of the set of paths. But I don't think that's the API that we should be exposing to 3rd party developers since it's a lot lower level. 

It seems like these permanent APIs are being added to solve temporary deficiencies in `core-data` which is unfortunate.

Have we considered preloading the different menus as separate requests?

anton-vlasenko commented:
@TimothyBJacobs 

> The structure of this PR is confusing to me. 

Yes, let's discuss it and maybe create follow-up tasks (given the age of this PR).

> Have we considered preloading the different menus as separate requests?

I'm not sure I understand.
Per @spacedmonkey's [comment](https://github.com/WordPress/gutenberg/issues/24642#issuecomment-924418340), the idea is to remove such calls.

> We should also remove call to __experimental/menus/<current_menu_id> as this data will be loaded in __experimental/menus

What is your idea exactly?

BTW, I've just realized we can get rid of that additional filter.
If we load the menu data needed for the `createMenuPreloadingMiddleware` function sideways, somewhere inside the `gutenberg_navigation_init` function, we don't need that additional filter.
We will have to call the code below again, but this way we can get rid of that additional filter.
```php
// Of course this snippet has to be refined, it's just for example
$preload_data = array_reduce(
		['/__experimental/menus'],
		'rest_preload_api_request',
		array()
	);
```

Would you please let me know if that would be an acceptable solution?

TimothyBJacobs commented:
> > Have we considered preloading the different menus as separate requests?
> 
> I'm not sure I understand. Per @spacedmonkey's [comment](https://github.com/WordPress/gutenberg/issues/24642#issuecomment-924418340), the idea is to remove such calls.
> 
> > We should also remove call to __experimental/menus/<current_menu_id> as this data will be loaded in __experimental/menus
> 
> What is your idea exactly?

Yeah, I don't have an issue with that `menus/<current>` route being preloaded even if we do also preload the menus collection. Though of course, preferably `core-data` should be able to just handle this. Is there a ticket for that?

> BTW, I've just realized we can get rid of that additional filter. If we load the menu data needed for the `createMenuPreloadingMiddleware` function sideways, somewhere inside the `gutenberg_navigation_init` function, we don't need that additional filter. We will have to call the code below again, but this way we can get rid of that additional filter.
> 
> ```
> // Of course this snippet has to be refined, it's just for example
> $preload_data = array_reduce(
> 		['/__experimental/menus'],
> 		'rest_preload_api_request',
> 		array()
> 	);
> ```
> 
> Would you please let me know if that would be an acceptable solution?

Yeah, not introducing that filter and implementing it as you suggest would be preferable to me.

anton-vlasenko commented:
>  Is there a ticket for that?

Yes, I created it several days ago. https://github.com/WordPress/gutenberg/issues/35601 (feel free to add your notes to it).
I'd like to explain it a bit.
When we load the navigation editor page, we make 2 duplicate calls to the `menus/<current>` endpoint.
And we don't make any calls to that endpoint when we change the currently selected menu.
So, I think the issue is somewhere inside the React components. It seems we don't need to call that `menus/<current>` at all. But we need to investigate it and understand if we can get rid of these duplicate calls.
We may not even need to refactor `core-data`.

> Yeah, not introducing that filter and implementing it as you suggest would be preferable to me.

Awesome! I will create a PR and an issue to refactor this.

anton-vlasenko commented:
@TimothyBJacobs 
> Yeah, not introducing that filter and implementing it as you suggest would be preferable to me.

So, here is the PR that removes that filter. I'd appreciate your review.
https://github.com/WordPress/gutenberg/pull/35838
